Improve highlighting by using highlight_query with all_fields enabled

lukasolson opened the pull request:
Fixes https://github.com/elastic/kibana/issues/2358.
Replaces https://github.com/elastic/kibana/pull/9091.

After [some discussion](https://github.com/elastic/kibana/pull/9091#issuecomment-267001252), we decided that the way to have the most reliable and predictable highlighting experience would be to utilize the `highlight_query` option, and use the same query as the request body query, with the exception that `all_fields` is enabled for any `query_string` queries.

This also adds another advanced setting, `doc_table:highlight:all_fields`, which defaults to `true`. If set to `false`, then the highlighting won't use a separate `highlight_query`.

This should resolve all of the cases that we discussed in https://github.com/elastic/kibana/pull/9091.

Bargs commented on src/ui/public/courier/data_source/_abstract.js:
```diff
@@ -389,6 +391,8 @@ export default function SourceAbstractFactory(Private, Promise, PromiseEmitter)
           });
         }(flatState.body.aggs || flatState.body.aggregations));
+
+        highlightBody(flatState.body);
```
I like that moving highlighting to _flatten ensures that it works across discover and dashboard. But it also means that even Vis requests ask for highlighting. I wonder if we should try to avoid this? Since the highlighting logic is nicely modularized now maybe we could just reuse it in both discover and dashboard?

<img width="1509" alt="screen shot 2016-12-29 at 12 56 24 pm" src="https://cloud.githubusercontent.com/assets/6239176/21550323/d0ddd874-cdc6-11e6-803e-c90d689843f4.png">

Bargs replied:
I also just realized this would overwrite any highlighting defined via the SearchSource.

weltenwort replied:
I would definitely vote for not adding more magic to the data source. This piece of code has cost me so much time when writing the log context view before giving up on it. The more we add the more we restrict its reuse.

lukasolson commented on src/ui/public/courier/data_source/_highlight_body.js:
```diff
@@ -0,0 +1,45 @@
+  if (_.has(clone, 'query_string')) {
+    clone.query_string.all_fields = true;
+  } else if (_.has(clone, 'bool.must')) {
+    clone.bool.must = clone.bool.must.map(getHighlightQuery);
```
That's kinda tricky, because what if the user has a field named `query_string` and is using that field inside a query? Then we might be adding `all_fields` in a place where it's not even allowed.

Bargs replied:
If only we had some kind of intermediate representation to avoid such issues... :)

It's a good point, I'll try to think on it over the weekend.

weltenwort replied:
I don't see a way to reliably do this without transforming the query into a normal form first either (which would be equivalent to an "intermediate representation"). :frowning_face:

tbragin commented:
Awesome progress! Functionally, highlighting works as I expect now, based on examples here https://github.com/elastic/kibana/pull/9091#issuecomment-261811764

The only weirdness I noticed was a refresh issue with whether the expanded document view for one of the documents in the preview was open during a search or not. If it was, there doesn't seem to be a refresh in terms of highlighting. 

![highlighting_sticky](https://cloud.githubusercontent.com/assets/933397/21632146/1ebfb960-d1f8-11e6-8a17-02318ff12820.gif)

lukasolson commented:
@tbragin Does that happen in master as well, or is it limited to this PR?

tbragin commented:
@lukasolson This issue seems to exists in master. Should I open a separate issue for it?

tbragin commented:
Went ahead and filed a separate issue https://github.com/elastic/kibana/issues/9729

lukasolson commented on src/ui/public/courier/data_source/_highlight_body.js:
```diff
@@ -0,0 +1,45 @@
+    fields: {
+      '*': {}
+    },
+    fragment_size: FRAGMENT_SIZE
```
Whoa, I never saw this comment for some reason. You hit the nail on the head... I wouldn't expect someone to turn this off unless they had a default field set. What do you think?

Bargs replied:
I agree, I also can't think of a reason to turn it off unless someone wants to search a specific field by default and in that case they probably only want highlighting on that specific field. I was probably worrying too much about preserving the old behavior :)

lukasolson commented:
Okay, this is ready for review again. A couple of notes:

* I'm not completely satisfied with the current implementation in that [`highlightRequest`](https://github.com/elastic/kibana/pull/9671/files#diff-d1695ba2026ff89878f9e4f4de683758R183) makes a call to `_flatten`. This means that any time we want to use this, `_flatten` will be called multiple times (once for highlighting and once to flatten for the actual request that gets sent). I'm open to suggestions here.
* I changed [`highlightTags`](https://github.com/elastic/kibana/pull/9671/files#diff-baf6a264c7e90958fbe595ddc2e30c71L7) from an Angular constant to a simple object export, and changed [`highlightFilter`](https://github.com/elastic/kibana/pull/9671/files#diff-6ef57afe1f02b1f3e8db3a2f14f0ba42L8) from an Angular filter to just a plain function ([`getHighlightHtml`](https://github.com/elastic/kibana/pull/9671/files#diff-b03542ccc5d3aaf4fbfbad15c54cce7bR6)). It wasn't being used anywhere in HTML anyway. I also added [an additional test](https://github.com/elastic/kibana/pull/9671/files#diff-3b42cd2936810ad816d7c5a42ff8c642R19) because it seems this was a condition that wasn't being tested.
* I updated the [`getHighlightQuery` logic](https://github.com/elastic/kibana/pull/9671/files#diff-bf2f2ea877513a00aebcd9fd29435accR9) to fix the bug that @Bargs mentioned above (when `must` is not an array).

Could you take another look, @Bargs and @weltenwort?

lukasolson commented on src/ui/public/highlight/highlight_html.js:
```diff
@@ -0,0 +1,29 @@
+
+export default function getHighlightHtml(fieldValue, highlights) {
+  let highlightHtml = (typeof fieldValue === 'object')
+    ? angular.toJson(fieldValue)
```
I can't remember exactly why this was put in here, but if I remember correctly, `fieldValue` could be an Angular object (with `$$`-prefixed values) and that's why we used `angular.toJson`.

Bargs replied:
Fair enough, not a big deal

Bargs commented on docs/management/advanced-options.asciidoc:
```diff
@@ -35,6 +35,8 @@ document.
 `doc_table:highlight`:: Highlight results in Discover and Saved Searches Dashboard. Highlighting makes request slow when
 working on big documents. Set this property to `false` to disable highlighting.
+`doc_table:highlight:all_fields`:: Improves highlighting by using a separate `highlight_query` that uses `all_fields` mode on
+`query_string` queries. Set to `false` if you are using a `default_field" in your index.
```
I think the double quote after default_field needs to be a backtick?

lukasolson commented:
Should this *only* apply to 5.x, seeing as how in 6.x the default will be to disable _all?

Bargs commented:
I think old indices will still have _all though right? I don't think _all will be completely removed until 7.0.

weltenwort commented on src/ui/public/courier/data_source/search_source.js:
```diff
@@ -178,6 +181,11 @@ export default function SearchSourceFactory(Promise, Private, config) {
     });
   };
 
+  SearchSource.prototype.highlightRequest = function highlightRequest() {
```
I'm probably missing something here, but where is this ever called?

lukasolson replied:
Oops, good point, that should have been removed.

Bargs commented:
🎆 🎉 🎈 !

lukasolson commented:
Backported to 5.x in 18451e5.
